src/tests/ftest/util/apricot/apricot/test.py

```diff
@@ -233,17 +247,65 @@ def setUp(self):
 
     def tearDown(self):
         """Tear down after each test case."""
-        try:
-            if self.agent_sessions:
-                self.d_log.info("Stopping agents")
-                agent_utils.stop_agent(self.agent_sessions,
-                                       self.hostlist_clients)
-        finally:
-            self.d_log.info("Stopping servers")
+        # Report any errors after attempting each tear down step
+        errors = []
+
+        # Destroy any containers first
+        if self.container:
+            self.dual_log("Destroying containers")
+            if not isinstance(self.container, (list, tuple)):
+                self.container = [self.container]
+            for container in self.container:
+                try:
+                    container.close()
+                    container.destroy()
+                except (DaosApiError, TestFail) as error:
+                    self.dual_log("  {}".format(error))
+                    errors.append(
+                        "Error destroying container: {}".format(error))
+
+        # Destroy any pools next
+        if self.pool:
+            self.dual_log("Destroying pools")
+            if not isinstance(self.pool, (list, tuple)):
+                self.pool = [self.pool]
+            for pool in self.pool:
+                try:
+                    pool.disconnect()
+                    pool.destroy()
+                except (DaosApiError, TestFail) as error:
+                    self.dual_log("  {}".format(error))
+                    errors.append("Error destroying pool: {}".format(error))
+
+        # Stop the agents
+        if self.agent_sessions:
+            self.dual_log("Stopping agents")
             try:
-                server_utils.stop_server(hosts=self.hostlist_servers)
-            finally:
-                super(TestWithServers, self).tearDown()
+                agent_utils.stop_agent(
+                    self.agent_sessions, self.hostlist_clients)
+            except agent_utils.AgentFailed as error:
+                self.dual_log("  {}".format(error))
+                errors.append("Error stopping agents: {}".format(error))
+
+        # Stop the servers
+        self.dual_log("Stopping servers")
+        try:
+            server_utils.stop_server(hosts=self.hostlist_servers)
+        except server_utils.ServerFailed as error:
+            self.dual_log("  {}".format(error))
+            errors.append("Error stopping servers: {}".format(error))
+
+        # Disable fault injection, if enabled
+        try:
+            super(TestWithServers, self).tearDown()
+        except OSError as error:
+            errors.append("Error disabling fault injection: {}".format(error))
+
+        # Fail the test if any errors occurred during tear down
+        if errors:
+            self.fail(
+                "Errors detected during teardown:\n  - {}".format(
+                    "\n  - ".join(errors)))
```

Comment:
Notice that the Functional stage at https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-1004/10/pipeline is FAILURE (red) rather than UNSTABLE (yellow)?  This is what we are trying to rectify.

Reply:
Both launch.py and ftest.sh yield a 0 return code in this case:
```
21:05:43  Detected one or more avocado test failures!
21:05:43  + rc=0
21:05:43  + unlink /var/lib/jenkins/jenkins-1/docker_1/workspace/daos-stack_daos_PR-1004/src/tests/ftest/avocado/job-results/latest
21:05:43  + ls 'core.*'
21:05:43  ls: cannot access core.*: No such file or directory
21:05:43  + exit 0
21:05:43  + false
21:05:43  + rc=0
21:05:43  + exit 0
```
Shouldn't this indicate that the stage completed successfully?  Is there something else that is determining that this stage is FAILED rather than UNSTABLE?

Reply:
> Shouldn't this indicate that the stage completed successfully?

At this point, yes, but there is more to the story...

> Is there something else that is determining that this stage is FAILED rather than UNSTABLE?

https://github.com/daos-stack/daos/blob/d9c91e4f76bcb18a217073b6a1acb5b6182d56df/Jenkinsfile#L1171

which results in https://github.com/daos-stack/pipeline-lib/blob/741a84b7dfd975e66895a5441871bdcf3dbc8d7e/vars/runTest.groovy#L84-L93